pkg/controller/job/job_controller.go

```diff
@@ -0,0 +1,297 @@
+		go wait.Until(v.worker, v.workerLoopPeriod, stopCh)
+	}
+
+	<-stopCh
+}
+
+func (v *JobController) enqueueJob(obj interface{}) {
+	key, err := controller.KeyFunc(obj)
+	if err != nil {
+		utilruntime.HandleError(fmt.Errorf("couldn't get key for object %+v: %v", obj, err))
+		return
+	}
+	v.queue.Add(key)
+}
+
+func (v *JobController) worker() {
+	for v.processNextWorkItem() {
+
+	}
+}
+
+func (v *JobController) processNextWorkItem() bool {
+	eKey, quit := v.queue.Get()
+	if quit {
+		return false
+	}
+
+	defer v.queue.Done(eKey)
+
+	err := v.syncJob(eKey.(string))
+	v.handleErr(err, eKey)
+
+	return true
+}
+
+// main function of the reconcile for job
+// job's name is same with the service that created it
+func (v *JobController) syncJob(key string) error {
+	startTime := time.Now()
+	defer func() {
+		log.V(4).Info("Finished syncing job.", "key", key, "duration", time.Since(startTime))
+	}()
+
+	namespace, name, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		return err
+	}
+
+	job, err := v.jobLister.Jobs(namespace).Get(name)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			return err
+		}
+		log.Error(err, "get job failed", "namespace", namespace, "name", name)
+		return nil
+	}
+
+	err = v.makeRevision(job)
+
+	if err != nil {
+		log.Error(err, "make job revision failed", "namespace", namespace, "name", name)
+		return nil
+	}
+
+	return nil
+}
+
+// When a job is added, figure out which service it will be used
+// and enqueue it. obj must have *batchv1.Job type
+func (v *JobController) addJob(obj interface{}) {
+	deploy := obj.(*batchv1.Job)
+
+	v.queue.Add(deploy.Name)
+
+	return
+}
+
+func (v *JobController) handleErr(err error, key interface{}) {
+	if err != nil {
+		v.queue.Forget(key)
+		return
+	}
+
+	if v.queue.NumRequeues(key) < maxRetries {
+		log.V(2).Info("Error syncing virtualservice for service, retrying.", "key", key, "error", err)
+		v.queue.AddRateLimited(key)
+		return
+	}
+
+	log.V(4).Info("Dropping service out of the queue", "key", key, "error", err)
+	v.queue.Forget(key)
+	utilruntime.HandleError(err)
+}
+
+func (v *JobController) makeRevision(job *batchv1.Job) error {
+	revisionIndex := -1
+	revisions, err := v.getRevisions(job)
+
+	// failed get revisions
+	if err != nil {
+		return nil
+	}
+
+	uid := job.UID
+	for index, revision := range revisions {
+		if revision.Uid == string(uid) {
+			currentRevision := v.getCurrentRevision(job)
+			if reflect.DeepEqual(currentRevision, revision) {
+				return nil
+			} else {
+				revisionIndex = index
+				break
+			}
+		}
+	}
+
+	if revisionIndex == -1 {
+		revisionIndex = len(revisions) + 1
+	}
+
+	revisions[revisionIndex] = v.getCurrentRevision(job)
+
+	revisionsByte, err := json.Marshal(revisions)
+	if err != nil {
+		glog.Error(err)
+	}
+
+	if job.Annotations == nil {
+		job.Annotations = make(map[string]string)
+	}
+
+	job.Annotations[revisionsAnnotationKey] = string(revisionsByte)
+	_, err = v.client.BatchV1().Jobs(job.Namespace).Update(job)
```

Comment:
have you go through k8s jobController, what will happen when you update a job, I think it trigger k8s jobController to run. better double check source code, in case of job rerun when you update